Documentation/proposals/202304-silences.md

```diff
@@ -0,0 +1,136 @@
+* Teams need to manage expiration on their own
+* If a silence was deleted by accident, it will be permanent.
+
+# Goals
+
+* Provide a way for users to manage Alertmanager silences with a Custom Resource Definition (CRD).
+* Enable management of Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...
+* Multi tenancy: Silences CR can be using kubernetes rbac.
+* Moving the Giant Swarm silence operator to prometheus operator would provide better community support and less fragmentation.
+* Improve user experience when managing silences, so users can use the same tools to manage their monitoring pipeline (kyverno, cr validations).
+
+## Audience
+
+* Users who use the Prometheus operator to offer monitoring-as-a-service to developers.
+* Users who want to manage silences the same way as for services running within the Kubernetes cluster.
+* Users who want a supported GitOps way of silences outside the Kubernetes cluster
+
+# Non-Goals
+
+* This proposal do not cover automatic cleanup of expired Silence CRs
+* Refactoring of the other CRDs is not in scope for the first version
+* Adding silences to a remote Alertmanager cluster
+
+# How
+
+Creating a new namespaced Silence CRD that will act as an interface by adding silences via the Alertmanager API.
+
+A typical `Silence` resource could look like the following:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: Silence
+metadata:
+  name: my-silence
+  namespace: monitoring
+  labels:
+    test: value
+spec:
+  expiresAt: "2023-06-01 10:00"
+  matchers:
+    - matchType: "="
+      name: namespace
+      value: test-ns
+```
+
+The above resource will result in creating a silence in the Alertmanager that has the appropriate selector.
+
+* `expiresAt` to define the expiration of the silence.
+  - If a silence has `expiredAt` defined, it'll expire at the time defined and the CR will have to be manually removed.
+  - If a silence doesn't have `expiredAt`, it'll be kept active as long as the CR exists and removed as soon as the CR is gone
+* `matchers` field corresponds to the Alertmanager silence matchers each of which consists of:
+  - `name` - name of tag on an alert to match
+  - `value` - fixed string or expression to match against the value of the tag named by name above on an alert
+  - `matchType` - Match operation available with AlertManager >= v0.22.0 and takes precedence over Regex (deprecated) if non-empty.
+  - `regex` - Whether to match on equality (false) or regular-expression (true). Deprecated as of AlertManager >= v0.22.0 where a user should use MatchType instead.
+
+This example doesn't list all the fields that are offered by Alertmanager. The implementation of all the fields will be
+done in an iterative process and as such, the expectation is not for all of them to be implemented in the first version.
+
+Also, to help select `Silences`, a new field will be added to the Alertmanager CRD:
+
+```yaml
+[...]
+spec:
+  silenceSelector: ...
```

Comment:
If this is selecting silences by labels, is it better to call it silenceLabelSelector ?